content/token22-in-the-client.md

```diff
@@ -0,0 +1,470 @@
+
+Lastly, `index.ts` contains our main script. It currently only creates a connection and calls `initializeKeypair` to generate the keypair for `payer`.
+
+
+### 3. Run validator node
+
+We suggest running your own validator for this lab (and most development/testing) to get around airdrop restrictions on devnet.
+
+In a separate terminal, run the following command: `solana-test-validator`. This will run the node and also log out some keys and values. The value we need to retrieve and use in our connection is the JSON RPC URL, which in this case is `http://127.0.0.1:8899`. We then use that in the connection to specify to use the local RPC URL.
+
+```tsx
+const connection = new Connection('http://127.0.0.1:8899', 'confirmed');
+```
+
+If you'd like to use Devnet and provide you're own devnet wallet, you still can - just reconfigure the `Connection` and the keypair path input to `initializeKeypair`.
+
+```tsx
+const connection = new Connection(clusterApiUrl('devnet'), 'confirmed');
+```
+
+Lets test that it all works so far by running `npm run start`. You should see the `payer` public key logged out in your terminal.
+
+
+### 4. Create Token Program and Token Extension Program mints
+
+Let's start by creating new token mints using both the `Token Program` and the `Token Extension Program`.
+
+Create a new file called `create-and-mint-token.ts`. In this file we will create a function called `createAndMintToken`. As the name suggests it will create a mint, token account (ATA) and then mint some amount of tokens to that account.
+
+Within this `createAndMintToken` function we will be call `createMint`, `getOrCreateAssociatedTokenAccount` and `mintTo`. This function is designed to be indifferent to the specific token program being used, allowing for the creation of tokens from either the `Token Program` or the `Token Extension Program`. This capability is achieved by accepting a program ID as a parameter, enabling the function to adapt its behavior based on the provided ID.
+
+Here are the arguments we'll be passing into this function:
+
+- `connection` - The connection object to use
+- `tokenProgramId` - The token program to point to
+- `payer` - The keypair paying for the transaction
+- `decimals` - The number of decimals to include for the mint
+- `mintAmount` - The amount of tokens to mint to the payer
+
+And this is what the function will do:
+
+- Create a new mint using **`createMint`**
+- Fetch mint information using **`getMint`**
+- Log mint information using **`printTableData`**
+- Create an associated token account with **`getOrCreateAssociatedTokenAccount`**
+- Log the address of the associated token account
+- Mint tokens to the associated token account with **`mintTo`**
+
+All put together this is what the final `createAndMintToken` function looks like:
```

Comment:
This is all super clear (what we will do, what functions do what, etc). Well done.